### PR TITLE
Add deprivilege/print/reprivilege flow

### PR DESCRIPTION
**Pull request by romkuz01**
Major changes:
- Rename old debug_deprivilege_and_return() to debug_deprivilege_and_die()
  which better reflects its nature
- Add a new debug_deprivilege_and_return() which indeed
  performs deprivilege and return within the function itself
- Add a new API function uvisor_debug_print_enable() which should
  be used from the user's application to enable
  deprivilege/print/reprivilege flow
- Make SVC's priority higher than of the other exceptions with programmable priority
- Normalize line endings used in printing in a few places to "\r\n" (still need to fix in some other places)

Note:
  Ideally it would be better to break this commit into a few smaller ones

Usage example in the user's application:
...
static void halt_error(THaltError reason, const THaltInfo * halt_info)
{
...
}

static void debug_print(const char * message)
{
	printf("%s", message);
}

UVISOR_PUBLIC_BOX_DEBUG_DRIVER(halt_error, debug_print);

#define BAD_BAD_ADDR (*((volatile unsigned long *) (0xFFFFFFFF)))

int main(void)
{
	uvisor_debug_print_enable(true);
 
   	/* About to get a fault now: */
        BAD_BAD_ADDR = 13;
}

### CODE REVIEW

File: core/debug/src/debug.c [hunk 1]
```diff
@@ -24,12 +24,12 @@
 #include <stdbool.h>
 
 #define DEBUG_PRINT_HEAD(x) {\
-    DPRINTF("\n***********************************************************\n");\
-    DPRINTF("                    "x"\n");\
-    DPRINTF("***********************************************************\n\n");\
+    DPRINTF("\r\n***********************************************************\r\n");\
```
Comment (alzix) [inline core/debug/src/debug.c]: please remove all \r and configure your serial terminal to force CR on each LF - it is a terminal configuration and should not be treated in code.

File: api/inc/api.h [hunk 2]
```diff
@@ -90,6 +90,7 @@ typedef struct {
     void (*spin_unlock)(UvisorSpinlock * spinlock);
 
     void (*debug_semihosting_enable)(void);
+    void (*debug_print_enable)(bool);
```
Comment (alzix) [inline api/inc/api.h]: this function comes to solve initialization order problems when using debug driver.
I believe this logic should reside on application side. An alternative is to look for any falg/callback from mbed-os - for checking such a condition.

Application can define a global `bool debug_print_enable = false;` variable, and avoid printing if it is disabled.
as i see it:
```
bool debug_print_enable = false;
void debug_print_driver(THaltError error_code, const THaltInfo *bsod)
{
    if (debug_print_enable){
        printf("oops\n");
    }
}
int main(void)
{
    debug_print_enable = true;

    ...

}

File: core/debug/src/debug.c [hunk 3]
```diff
@@ -50,25 +50,31 @@ void debug_semihosting_enable(void)
 
 UVISOR_WEAK void default_putc(uint8_t data)
 {
-    if (DEBUG_SEMIHOSTING_MAGIC == g_semihosting_magic) {
```
Comment (alzix) [inline core/debug/src/debug.c]: i suggest not to remove semi-hosting print logic - we still need it for debugging stuff during early boot stages.
we can duplicate print to two sinks - that is what all decent loggers do :)

File: core/system/src/core_armv7m/virq.c [hunk 4]
```diff
@@ -519,17 +540,17 @@ void virq_init(uint32_t const * const user_vtor)
      * levels. */
     assert(__UVISOR_NVIC_MIN_PRIORITY < UVISOR_VIRQ_MAX_PRIORITY);
 
-    /* Set the priority of each exception. SVC is lower priority than
-     * MemManage, BusFault, UsageFault, and SecureFault_IRQn, so that we can
-     * recover from security violations more simply. */
+    /* Set the priority of each exception. SVC is set to higher priority than
```
Comment (Patater) [inline core/system/src/core_armv7m/virq.c]: Please change the priority of exceptions in its own commit, so we can test that changing the priority doesn't cause us to fail tests, independently of other changes you've made.

File: core/vmpu/src/mpu_armv7m/vmpu_armv7m.c [hunk 5]
```diff
@@ -137,6 +137,17 @@ uint32_t vmpu_sys_mux_handler(uint32_t lr, uint32_t msp)
         halt_info = &info;
     }
 
+    /* If we have a Hard Fault due to the Bus or MemManage Fault escalation we'll try to treat
+     * it as the original fault. */
```
Comment (Patater) [inline core/vmpu/src/mpu_armv7m/vmpu_armv7m.c]: The handling of escalated hard faults should also be in its own commit, so we can test it independently of other changes.

What happens if we hard fault a second time as a result of trying to handle the escalated fault? Do we enter a hard fault forever loop? Such a situation seems not so great.

File: core/vmpu/src/mpu_armv7m/vmpu_armv7m.c [hunk 6]
```diff
@@ -137,6 +137,17 @@ uint32_t vmpu_sys_mux_handler(uint32_t lr, uint32_t msp)
...
+    if (ipsr == HardFault_IRQn && (SCB->HFSR & SCB_HFSR_FORCED_Msk)) {
+    	if (VMPU_SCB_BFSR) {
+    		ipsr = BusFault_IRQn;
+    	} else
+    	if (VMPU_SCB_MMFSR) {
```
Comment (Patater) [inline core/vmpu/src/mpu_armv7m/vmpu_armv7m.c]: Please follow style guidelines. `else` and `if` should be on same line here.

File: core/system/src/core_armv7m/virq.c [hunk 7]
```diff
@@ -324,6 +324,27 @@ int virq_irq_level_get(void)
...
+    );
+}
+
+/* Restore the execution right after the point of deprivilege, the needed stack frame
+ * was created on MSP by SVC used for deprivileging. */
+void UVISOR_NAKED virq_gateway_return(uint32_t svc_sp, uint32_t svc_pc)
+{
+    /* Return to handler mode with MSP using a basic stack frame. */
+    asm volatile(
+        "ldr lr, =0xfffffff1\n"
```
Comment (Patater) [inline core/system/src/core_armv7m/virq.c]: Would anything bad happen if another uVisor SVC was called before the`virq_gateway_return` SVC? What would happen to MSP? Are we taking any steps to prevent calls to other SVC while in the debug box? Can the `virq_gateway_return` SVC be used to leak information stored on the MSP to its caller?

File: core/debug/src/debug.c (hunk 3 above)
Comment (romkuz01) [reply core/debug/src/debug.c]: Not removed, divided into two separate checks.

File: core/debug/src/core_armv7m/debug_box_armv7m.c [hunk 8]
```diff
@@ -68,3 +68,50 @@ void debug_deprivilege_and_die(void * debug_handler, void * return_handler,
...
+    ((uint32_t *)dst_sp)[1] = a1;
+    ((uint32_t *)dst_sp)[2] = a2;
+    ((uint32_t *)dst_sp)[3] = a3;
+
+    context_switch_in(CONTEXT_SWITCH_FUNCTION_DEBUG, dst_id, __get_PSP(), dst_sp);
+
+    /* Save the current context on the stack, deprivilege and restore the context upon return. */
+    asm volatile(
+        /* Save general purpose registers that won't be saved by the following SVC. */
+        "push {r4 - r11}\n"
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/core_armv7m/debug_box_armv7m.c]: No need since LR is automatically saved when executing SVC after this PUSH.
Also R0-R3, R12 and xPSR.

File: core/debug/src/core_armv8m/debug_box_armv8m.c [hunk 9]
```diff
@@ -37,7 +37,21 @@ void debug_deprivilege_and_die(void * debug_handler, void * return_handler,
...
+    /* FIXME: the below way of deprivileging may be problematic when executed from an exception handler
+     * since we're going to stay in the context of the exception with IPSR reflecting that.
+     * We need to do deprivileging in a way similar to ARMv7 when an exception frame is forged
+     * for that purpose. */
     uint32_t caller = UVISOR_GET_NS_ALIAS(UVISOR_GET_NS_ADDRESS((uint32_t) debug_handler));
     SECURE_TRANSITION_S_TO_NS(caller, a0, a1, a2, a3);
     ((void (*)(void)) return_handler)();
 }
+
+/* FIXME: replace these stubs by the actual implementation. */
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/core_armv8m/debug_box_armv8m.c]: Probably it's better to do it in a separate PR.
This one is already big enough.

File: core/debug/src/core_armv7m/debug_box_armv7m.c [hunk 10]
```diff
@@ -68,3 +68,50 @@ void debug_deprivilege_and_die(void * debug_handler, void * return_handler,
...
+        
+        /* Execute SVC that will perform the deprivileging. */
+    	"svc %[depriv]\n"
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/core_armv7m/debug_box_armv7m.c]: Right now we can't perform deprivileging from SVC handler since executing SVC in this case will result in Hard Fault.
We need either disable deprivileging from SVC like it's done for Hard Fault/NMI or manually play with active bit. I'll check this option.

File: core/debug/src/debug_box.c [hunk 11]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
 
 uint32_t g_debug_interrupt_sp[UVISOR_MAX_BOXES];
 
+void debug_print(const uint8_t * message_buffer, uint32_t size)
+{
+	/* Abort printing if debug box wasn't initialized yet. */
+	if (!g_debug_box.initialized) {
+		return;
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Currently the initialization header isn't printed out:

uVisor mode: 2
uvisor_ram : @0x1FFF0400 (7168 bytes) [config]
             @0x1FFF0400 (7168 bytes) [linker]
bss_boxes  : @0x1FFF2000 (128 bytes) [config]
Box 0 ACLs:
  - Peripheral: 0x40047000 - 0x40048064 (permissions: 0x0AB6)
  - Peripheral: 0x40065000 - 0x40065001 (permissions: 0x0AB6)
  - Peripheral: 0x40064000 - 0x4006400E (permissions: 0x0AB6)
  - Peripheral: 0x40049000 - 0x400490CC (permissions: 0x0AB6)
  - Peripheral: 0x4004A000 - 0x4004A0CC (permissions: 0x0AB6)
  - Peripheral: 0x4004B000 - 0x4004B0CC (permissions: 0x0AB6)
  - Peripheral: 0x4004C000 - 0x4004C0CC (permissions: 0x0AB6)
  - Peripheral: 0x4004D000 - 0x4004D0CC (permissions: 0x0AB6)
  - Peripheral: 0x4003D000 - 0x4003D808 (permissions: 0x0AB6)
  - Peripheral: 0x40040000 - 0x40040010 (permissions: 0x0AB6)
  - Peripheral: 0x40037000 - 0x40037140 (permissions: 0x0AB6)
  - Peripheral: 0x4007E000 - 0x4007E004 (permissions: 0x0AB6)
  - Peripheral: 0x4006A000 - 0x4006A020 (permissions: 0x0AB6)
  - Peripheral: 0x40066000 - 0x4006600C (permissions: 0x0AB6)
  - Peripheral: 0x4002C000 - 0x4002C08C (permissions: 0x0AB6)
vmpu_enumerate_boxes [DONE]
page heap: [0x20000000, 0x2000a000] 40kB -> 5 8kB pages
uvisor initialized

Debug box becomes initialized right before "vmpu_enumerate_boxes [DONE]":
    ...
        g_debug_box.initialized = 1;
    }

    /* Load box 0. */
    context_switch_in(CONTEXT_SWITCH_UNBOUND_FIRST, 0, 0, 0);

    DPRINTF("vmpu_enumerate_boxes [DONE]\n");

However interrupts are still disabled at this point which is problematic for deprivilege/reprivilege flow.
The next two lines are still printed with the interrupts disabled, therefore deprivileging isn't performed for them.

File: core/debug/src/debug_box.c [hunk 12]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	}
+
+	/* Debug print relies on "deprivilege and return" flow that will currently
+	 * only work if:
+	 * - executed with exceptions enabled
+     * - executed from within the exception context, i.e. when IPSR isn't 0
+     * - the current exception priority is lower than of Hard Fault
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Converted tabs inserted by mistake into spaces.

File: core/debug/src/debug_box.c [hunk 13]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	 * These limitation are dictated by the fact that SVC is used for
+	 * deprivileging and in order for it to work SVC must be executed with
+	 * exceptions enabled and its priority must be higher than of the current exception.
+	 * SVC's priority is set to be higher than of any other exception with
+	 * programmable priority but this still can't beat NMI and Hard Fault.
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: I'll try to make the lines as long as possible, but if a line is ended with a long word that crosses the boundary it has to be moved to the next line making the original line significantly shorter.

File: core/debug/src/debug_box.c [hunk 14]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	 */
+
+	/* Make sure we're within an exception and exception are enabled. */
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Will fix the typo.

File: core/debug/src/debug_box.c [hunk 15]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	int exception = (int)(__get_IPSR() & 0x1FF);
+	if (!exception || __get_PRIMASK() || __get_FAULTMASK())	{
+		return;
+	}
+
+	/* Bring exception number to CMSIS format and check exception's priority. */
+	exception -= NVIC_OFFSET;
+	if (exception == NonMaskableInt_IRQn || exception == HardFault_IRQn) {
+		return;
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Converted tabs inserted by mistake into spaces.

File: core/debug/src/debug_box.c [hunk 16]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	}
+
+	/* Place the message on the interrupts stack allocating
+	 * the size rounded up to the next value divisible by 4. */
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Changed the comment to:

/* Place the message on the interrupt stack allocating the size rounded up
 * to the next value divisible by 4. This way the stack stays aligned on
 * word boundary. */

File: core/debug/src/debug_box.c [hunk 17]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	uint32_t sp = g_debug_interrupt_sp[g_debug_box.box_id];
+	g_debug_interrupt_sp[g_debug_box.box_id] -= ((size + 3) / 4) * 4;
+	void *message = (void *)g_debug_interrupt_sp[g_debug_box.box_id];
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Not sure where exactly you propose to use sp.
If in line
void *message = (void *)g_debug_interrupt_sp[g_debug_box.box_id];
then sp without manipulations won't be good since we just updated g_debug_interrupt_sp[g_debug_box.box_id] and sp contains the original value.

File: core/system/src/core_armv7m/virq.c (hunk 4 above)
Comment (romkuz01) [reply core/system/src/core_armv7m/virq.c]: Made it a separate commit in this PR.

File: core/vmpu/src/mpu_armv7m/vmpu_armv7m.c (hunk 5 above)
Comment (romkuz01) [reply core/vmpu/src/mpu_armv7m/vmpu_armv7m.c]: Made it a separate commit in this PR.

Still need to think whether we can stuck in a Hard Fault loop.

File: core/debug/src/core_armv7m/debug_box_armv7m.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/core_armv7m/debug_box_armv7m.c]: OK

File: core/debug/src/core_armv8m/debug_box_armv8m.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/core_armv8m/debug_box_armv8m.c]: Can we remove these empty stubs or does that break the build?

File: core/debug/src/debug_box.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/debug_box.c]: OK

File: core/debug/src/debug_box.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/debug_box.c]: We currently share buffers for semihosting and debug box. If we made the two buffers separate, could we cache the debug box buffer output until the debug box is usable? That way, we could get these early prints. Not sure if we want two buffers, though. I'm open to other ideas to get these early prints.

File: core/debug/src/debug_box.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/debug_box.c]: Yup, that's right. Many text editors can automatically wrap the paragraph for you. Consider using a text editor that can do this.

File: core/debug/src/core_armv8m/debug_box_armv8m.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/core_armv8m/debug_box_armv8m.c]: Breaks the build.

File: core/debug/src/debug_box.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/debug_box.c]: Why does the stack need to be aligned on 4-byte boundary? Sometimes it has to be aligned to 8-byte boundary, right? Is 4-byte alignment always enough here?

File: core/debug/src/debug_box.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/debug_box.c]: Couldn't you assign `message` before you modify `g_debug_interrupt_sp[g_debug_box.box_id]`?

File: core/system/src/core_armv7m/virq.c (hunk 4 above)
Comment (Patater) [reply core/system/src/core_armv7m/virq.c]: OK

Thanks

File: core/debug/src/debug_box.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: As far as I know 4-byte alignment is an architectural requirement - you can't even set the two lowest bits of SP to be non-zero. 8-byte alignment is enforced by the architecture by automatically inserting padding when creating and exception stack frame.

In our case 4-byte alignment is sufficient, context_forge_exc_sf() will add padding if necessary.

File: core/debug/src/debug_box.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Maybe I wasn't clear enough, but we do need to assign modified value of g_debug_interrupt_sp[g_debug_box.box_id] into message (the new stack top after reserving space for the message).
Still can't see a way to make the code prettier.

File: core/debug/src/debug.c (hunk 1 above)
Comment (romkuz01) [reply core/debug/src/debug.c]: Done.

File: core/debug/src/debug_box.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (alzix) [reply core/debug/src/debug_box.c]: AAPCS compliance requires double-word stack alignment.
Since we are calling functions implemented in C  - we better to be complaint with calling convention.
@Patater - nice catch!

**Comment by romkuz01** [comment]
retest uvisor

File: core/debug/src/core_armv7m/debug_box_armv7m.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/core_armv7m/debug_box_armv7m.c]: I added active bit cleaning, so now deprivileging from SVC handler is possible.

File: core/debug/src/debug_box.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: I'll think about the best way to do it.

File: core/debug/src/debug_box.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: I've changed the alignment to be 8 bytes.

File: core/debug/src/debug_box.c [hunk 18]
```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
...
+	memcpy(message, message_buffer, size);
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: Changed to memmove.

File: core/system/src/core_armv7m/svc.c [hunk 19]
```diff
@@ -135,7 +135,11 @@ void UVISOR_NAKED SVCall_IRQn_Handler(void)
         ".align 4\n"                                // the jump table must be aligned
     "jump_table_unpriv:\n"
         ".word  virq_gateway_out\n"
+#ifdef NDEBUG
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/system/src/core_armv7m/svc.c]: Removed the conditional compilation.

File: core/system/src/core_armv7m/virq.c [hunk 20]
```diff
@@ -324,6 +324,27 @@ int virq_irq_level_get(void)
     return (int) NVIC_GetPriority(irqn) - __UVISOR_NVIC_MIN_PRIORITY;
 }
 
+/* Jump to unprivileged thread mode, the stack frame is already prepared on PSP. */
+void UVISOR_NAKED virq_gateway_deprivilege(uint32_t svc_sp, uint32_t svc_pc)
```
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/system/src/core_armv7m/virq.c]: Renamed deprivilege and return functions to:
debug_uvisor_deprivilege
debug_uvisor_return
and moved them to debug_box.c

File: core/system/src/core_armv7m/virq.c (hunk 4 above)
Comment (romkuz01) [reply core/system/src/core_armv7m/virq.c]: Changed the design to reset active bits, so reverted this change.

File: core/vmpu/src/mpu_armv7m/vmpu_armv7m.c (hunk 5 above)
Comment (romkuz01) [reply core/vmpu/src/mpu_armv7m/vmpu_armv7m.c]: Changed the design to reset active bits, so reverted this change.

File: core/system/src/core_armv7m/virq.c (hunk 7 above)
Comment (romkuz01) [reply core/system/src/core_armv7m/virq.c]: It's possible to execute SVC during deprivileging since deprivileged code operates at lower exception level.
No information is supposed to leak during return - we're using the previously created on MSP stack frame.

File: core/debug/src/debug_box.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (romkuz01) [reply core/debug/src/debug_box.c]: The stack memory is managed by uVisor itself and is supposed to
be accessible by debug box, so there's no need for the access check.

I've added a comment saying that.

**Comment by romkuz01** [comment]
@mikisch81 @orenc17A @theamirocohen
Please review

File: core/debug/src/core_armv8m/debug_box_armv8m.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (Patater) [reply core/debug/src/core_armv8m/debug_box_armv8m.c]: OK

**Comment by Patater** [comment]
Merged debug box printing to dev/romkuz01/debug-box-print-experimental due to high risk change and lack of thorough testing

Merged ready to go commits to master
